modules/getComponents.js

```diff
@@ -1,15 +1,37 @@
 import { mapAsync } from './AsyncUtils'
+import warning from './routerWarning'
 
-function getComponentsForRoute(location, route, callback) {
+function getComponentsForRoute(nextState, route, callback) {
   if (route.component || route.components) {
     callback(null, route.component || route.components)
-  } else if (route.getComponent) {
-    route.getComponent(location, callback)
-  } else if (route.getComponents) {
-    route.getComponents(location, callback)
-  } else {
-    callback()
+    return
   }
+
+  const getComponent = route.getComponent || route.getComponents
+  if (getComponent) {
+    // By assumption, location is a POJSO, and can be spread into the router
+    // state. The merging means that we can't use deprecateObjectProperties.
+    const { location } = nextState
+    let nextStateWithLocation = { ...nextState, ...location }
+
+    if (__DEV__) {
+      if (typeof Proxy === 'function') {
+        nextStateWithLocation = new Proxy(nextStateWithLocation, {
```

Comment:
We can use something like `deprecateObjectProperties`, but it'd be a bit more code. This is what React does upstream to prevent people from messing with synthetic events: https://github.com/facebook/react/blob/v15.0.1/src/renderers/dom/client/syntheticEvents/SyntheticEvent.js#L188.